### DISCUSSION TITLE
Editing resources via StructureMaps

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#1687] [Ideas]
# Business Requirement

## Changing patient programs
- When registering patients into different programs e.g ANC & PNC we create a condition that is attached to these patients.
- When the patient is done with the program we need to deactivate the condition and in some cases add the patient to a new program e.g 
    - When an ANC patient has a pregnancy outcome they need to move to PNC of close ANC depending on the outcome. 
    - This means we need to change the ANC condition `[clinicalStatus](https://www.hl7.org/fhir/condition-definitions.html#Condition.clinicalStatus)` to `resolved` or `inactive`

## Updating Family details
- Currently, we use the Group resource to track the families on FHIR Core. This means the `member` attribute of the Group resource is used to hold the family member references. 
- We need to update the `member` attribute once every time we add a new family member or remove an existing family member. 

# Technical Requirement
- Be able to update resources through StructureMaps. This is to avoid having to write custom Kotlin code for such resource updates. 
- The idea is to find a way to achieve the following items to satisfy the editing resources via StructureMaps requirement. 
   - [ ] Fetch the resources to be edited and attach them to the questionnaire triggering the edit. 
   - [ ] Update the resources via rules on the StructureMap 
   - [ ] Save the edited resource as an edit. This requires some changes to the SDK. 

## Saving the extracted resources correctly

**Comment by pld** [comment]
Other than StructureMaps, are there any other ways to update resource without custom code?

**Reply by maimoonak** [reply]
We do have https://build.fhir.org/ig/HL7/sdc/extraction.html
- Observation based - only for Observation - not implemented yet in SDK though
- Definition based - the one we were doing for Patient in initial questionnaires like [this](https://github.com/opensrp/fhircore/blob/anc_module/android/eir/src/main/assets/sample_patient_registration.json#L10). 
  - This had issues with multiple type of entities mapping which can be fixed
  - Main problem was that it would make questionnaire extremely lengthy and complex when we have a lot of hidden fields mapping like coding, priority, service-type, owner, created-on, etc. 
  - It would mean that we do mapping of each and every field of entities with ALL fields with default/required/same values. For example in Observation/Encounter there are only 2/3 fields which are driven from questionnaire input but rest of the fields are added as per business logic with default/same values
  - This also makes it difficult to automate the questionnaire generation as product team does it rightnow. They would have to make extra effort to add and then keep it updated when editing questionnaires.
  - We still might not be able to do complex operations specifically looping on these fields
  - This is very handly for simple mappings though as we were doing with Patient data where almost all fields were extracted from an input field
- Structure map

**Comment by maimoonak** [comment]
**How to identify the linked resource and update it on the StructureMap**
The Bundle() in SDK ResourceMapper.extractByStructureMap is the entity which is output by parser with corresponding entities (Patient/Encounter/Obs ... etc) that is ready to be saved to database by caller. This Bundle is actually passed to StructureMap processor/transformer as well and is populated by relevant final data. If this Bundle is passed externally i.e. from caller, then a structure-map script can do the createOrUpdate operation. Rightnow this is always new thats why we always get new entities and can not keep track of existing ones into StructureMap
```
  private suspend fun extractByStructureMap(...): Bundle {
    val structureMapProvider = structureMapExtractionContext.structureMapProvider
    val simpleWorkerContext = DataCapture.getConfiguration..................
    val structureMap = structureMapProvider(questionnaire.targetStructureMap!!, simpleWorkerContext)

    return Bundle().apply {
      StructureMapUtilities(
          simpleWorkerContext,
          structureMapExtractionContext.transformSupportServices
        )
        .transform(simpleWorkerContext, questionnaireResponse, structureMap, this)
    }
  }
```

**Comment by maimoonak** [comment]
**How to correctly extract the edits via the extraction functionality on SDK**
The StructureMap script allows to filter resource by fhirpath as `bundle.entry.where(resource is Condition and resource.code.coding.code='1234').resource) `. 
If we are able to pass the existing data into given Bundle then we can filter the bundle entries by fhirpath based on required properties (OR create a new one if none exists). This would be similar to createOrUpdate operation as existing resource would be updated in case it exists
